unicef/etools
# Feature flip implementation

Connects unicef/etools-issues#599

Here is my WIP for feature-flipping. I need feedback on the approach/design so please review mercilessly!

We use the [django-waffle](https://github.com/jsocol/django-waffle) package and I created a new app called `features` to modify it slightly to allow segmentation of users by 'country'.

The UI for adding or changing a flag is via the Waffle admin. The important thing is to give it a name and assign it countries via the multiselect box at the bottom. There are multiple options on that page to assign specific users, groups, percentages, etc. All of that comes from Waffle. Which leads to my first question:

Should we allow admins to feature flip ONLY by country? We could start that way, simplifying both the UI to create flags and to check them. It wouldn't be hard to later incorporate more of django-waffles segmentation options (users, groups, percentages) or add our own (user/country combo, etc).
We could choose not to show any of those other fields, and only show countries, but I wanted to show you the possibilities. Alternatively, we could hide this from the admin and create our own UI for tag management.

Segmenting options:
1. Only segment by country
2. Allow segmentation by country plus all of waffle's options

UI options:
1. Use waffle admin, showing all possibilities
2. Use waffle admin, showing only countries.
3. Create our own admin, showing only countries.

Django-waffle also has the concept of switches (always on or off) and samples which are not
dependent on any info in the request. I haven't exposed those via the API, but you can see them in the admin. Flags can be used to accomplish the same thing as switches and samples, with perhaps a tiny performance benefit to the latter.

For the API, I went with Cory and Greg's advice to have a read-only endpoint that only returns the active
flags for a request.

To test this PR:
1. go to the bottom of the admin page and under Waffle, click 'Add Flag'
2. Create a flag named 'can_view_foo' and assign it a few countries (way at the bottom of the form). Make sure that you uncheck the superuser box otherwise the flag will always be on.
3. Go back to the dashboard (/dash/) and make sure you're in the workspace of a country that you added
to the flag.
4. Load /api/v2/features/flags/ -> should return JSON `{active_flags: ['can_view_foo']}`
5. Go back to the dashboard and choose a different country
6. Load /api/v2/features/flags/ -> should return JSON `{active_flags: []}`

Before I call this ready, I will add:
- [ ] Docs for admins on how to add flags
- [ ] Docs for devs on how to use flags, switches, and samples

What else?
